Revert to status branch mk.2

Damon3000s opened the pull request:
Re-branching due to incorrectly merging the wrong dev branch in.

This change will add a "Rever to Status Branch" option to the UEGitPlugin. The purpose is that you can select any file which has been pushed to Git and revert it back to what is in in the Status Branch (eg. Main) if you no longer wanted the changes anymore. We are adding it to the plugin for a significant usability improvement due to the current workflow being "download the file from the repo directly, lock the file, replace the file, commit and push the file"

Darcy3000s committed:
Add a button in the right click context menu on assets which allows you to "Revert to status branch" (Rollback to Origin) - notably this doesn't actually perform any reverts, it just pulls the latest version of that file from main into your branch

Darcy3000s committed:
Diff files before attempting to sync them from another branch so we only sync files which are actually different - bail if we try to sync a file that we don't have to ability to checkout (It's locked by somebody else or changed in a branch other than status branch). Don't try to checkout files which are already checked out

Darcy3000s committed:
Change icon for Status Revert to use Revert icon

Damon3000s committed:
First pass at fixing Revert To Status Branch

The issue is that paths are in different formats and comparisons don't work (see comments). This works for the base game but requires more testing for plugin content. Matt has also recommended to look into "CreateStandardFilename" and "MakeStandardFilename" as a better alternative, but I want to get this "working" version in first.

Damon3000s committed:
Improve naming and how we substring

Damon3000s committed:
Simplify path conversions

Converting DiffResults to absolute paths is easier than converting absolute paths to git repo relative paths. Less code, less problems.

Damon3000s committed:
Use ApplyOperationAndReloadPackages

This will cause the file to be unlinked, reloaded and relinked. This way Unreal won't know the file has changed and stomp it.

Damon3000s committed:
Add missing include

Damon3000s commented:
See here for the original branch and PR with comments (all resolved) https://github.com/TheThreeThousands/UEGitPlugin/pull/2

Damon3000s committed:
Don't display Revert To Status Branch for maps

Damon3000s committed:
Don't display the Revert To Status Branch option for Maps

The option doesn't work for maps, so better to not show it.

It should work for maps that are not One Actor Per File, if the correct options are passed to the reload function, but I haven't gone through the work to attempt to detect that situation.

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,92 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+
+		Provider.Execute(ISourceControlOperation::Create<FCheckOut>(), FilesToSync);
+
+		USourceControlHelpers::ApplyOperationAndReloadPackages(FilesToSync, [&](const TArray<FString>&)
```
Don't just take a ref to everything

Damon3000s replied:
Done

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,92 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+
+		USourceControlHelpers::ApplyOperationAndReloadPackages(FilesToSync, [&](const TArray<FString>&)
+		{
+			// Git checkout is pulling content
```
Better comment

Git checkout (download) != Unreal checkout (lock)

Damon3000s replied:
Done

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,92 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+			// Git checkout is pulling content
+			TArray<FString> Results;
+			TArray<FString> Errors;
+			RunCommand("checkout", InPathToGitBinary, InRepositoryRoot, { BranchName, TEXT("--") }, FilesToSync, Results, Errors);
```
Check return value and error list.

Damon3000s replied:
Done. Logging to console as it didn't seem to log anything when I introduced errors by using an invalid command

Damon3000s committed:
Explicitly capture variables in the Lambda

Damon3000s committed:
Update comment that tries to disambiguate Unreal checkout and git checkout

Damon3000s committed:
Only commit if the revert was a success

Also log the operation errors

Damon3000s committed:
Only unlock locked files

Darcy3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,112 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		}
+		else
+		{
+			const FText Message = LOCTEXT("FailedBranchSyncGitCheckout", "Revert To Status Branch failed -  git checkout failed. See the log for details");
```
We should run an FRevert if the operation failed so we don't leave people in a half updated state with some files checked out

Damon3000s replied:
I was unable to get reverting working in that specific location, but I could check for a failed `git checkout` within the `USourceControlHelpers::ApplyOperationAndReloadPackages` lambda and do a synchronous revert operation.

Damon3000s replied:
Undid the revert change as my implementation was reverting ALL files and is is cleaner code and safer to not assume what files we want to revert and get someone to check or ask for help.

The only time it will fail is if the git checkout step fails, which is hopefully rare.

gpotrykus committed:
bugfix: fix git diff command to compare between actual head instead of main status branch, as we already do compare against all status branches in the loop

mastercoms committed:
Merge pull request #213 from gpotrykus/bugfix_status_branch_comparison

bugfix: changes on status branches are not detected

TwentyPast4 committed:
UE 5.6 workaround

mastercoms committed:
Merge pull request #216 from Mjolkling-Games/dev

UE 5.6 workaround

Commit:
Fix UE 5.5 Build
Building on UE5.5 and probably earlier broken by the UE 5.6 workaround as it tries to call GLevelEditorModeTools().CanAutoSave() without the appropriate header.

mastercoms committed:
Merge pull request #218 from MichaelManley/fix-ue55

Fix UE 5.5 Build

Damon3000s committed:
Merge remote-tracking branch 'origin/dev' into revert_to_status_branch_clean_2

Damon3000s committed:
Make a bool constexpr

Damon3000s committed:
First pass of reverting files if 'sync to status branch' failed

Damon3000s committed:
Merge remote-tracking branch 'origin/revert_to_status_branch_clean_2' into revert_to_status_branch_clean_2

Damon3000s committed:
Optimise reverting files if they fail to sync to status branch

Instead of trying to revert everything that was selected, only try to revert what we tried to sync to status branch

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+			FCheckinResultInfo ResultInfo;
+			// Re-check the status of the files before opening the window because we've just reverted a bunch of files which haven't had a status update yet.
+			constexpr bool bUseSourceControlStateCache = false;
+			FSourceControlWindows::PromptForCheckin(ResultInfo, FilesToSync, TArray<FString>(), TArray<FString>(), bUseSourceControlStateCache);
```
Maybe we don't actually want this?

Damon3000s replied:
Done

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+
+		if (bOperationSuccess)
+		{
+			FCheckinResultInfo ResultInfo;
```
Display a confirmation box instead

Damon3000s replied:
Done

Damon3000s commented:
Possibly unhandled situations
* Modify an existing file, save, revert to status -> File will download what is on main, be locked but can't be committed (because it was never originally committed)
    * We should run Revert on those files instead
* What happens if you run this on a new file that has never been committed? Should we delete? Not allow the option.

Damon3000s commented on Source/GitSourceControl/Public/GitSourceControlUtils.h:
```diff
@@ -387,4 +387,6 @@ bool PullOrigin(const FString& InPathToGitBinary, const FString& InPathToReposit
 
 GITSOURCECONTROL_API TSharedPtr< class ISourceControlRevision, ESPMode::ThreadSafe > GetOriginRevisionOnBranch( const FString & InPathToGitBinary, const FString & InRepositoryRoot, const FString & InRelativeFileName, TArray< FString > & OutErrorMessages, const FString & BranchName );
 
+void SyncAssetsFromBranch(const FString& InPathToGitBinary, const FString& InRepositoryRoot, const TArray<FAssetData>& SelectedAssets, FString BranchName);
```
const ref

Damon3000s replied:
Done

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
 
+void SyncAssetsFromBranch(const FString& InPathToGitBinary, const FString& InRepositoryRoot, const TArray<FAssetData>& SelectedAssets, FString BranchName)
+{
+	TArray<FString> FilesToSync;
```
Can this name be better?

Damon3000s replied:
Renamed

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+	for (const FAssetData& AssetData : SelectedAssets)
+	{
+		FString PackageName = AssetData.PackageName.ToString();
+		FString AbsoluteFileName = SourceControlHelpers::PackageFilename(PackageName);
```
AbsoluteFilePath

Damon3000s replied:
Renamed

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+
+	TArray<FString> DiffResults;
+	TArray<FString> ErrorMessages;
+	TArray<FString> DiffParametersLog{ TEXT("--pretty="), TEXT("--name-only"), FString::Printf(TEXT("HEAD..%s"), *BranchName), TEXT(""), TEXT("--") };
```
Is HEAD..%s correct here? upstream changed how we diff things

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		Path = FPaths::ConvertRelativePathToFull(InRepositoryRoot, Path);
+	}
+
+	FilesToSync.RemoveAllSwap([&DiffResults](const FString& File)
```
Comment what/why this is doing

Damon3000s replied:
Can this just be FilesToSync = DiffResults?

Damon3000s replied:
Assignment is better and hopefully more clear

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,125 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+
+	// DiffResults is repo relative and we need to be
+	// in absolute so we can compare against FilesToSync
+	for (FString& Path : DiffResults)
```
Could this be FilesToSync.Clear. Resize(DiffResults.Num) for (...) FilesToSync.Add(x)

Damon3000s replied:
FilesToSync = DiffResults is cleaner

Damon3000s committed:
Pass parameter by const ref

Damon3000s committed:
Replace checkin prompt with success message box

Damon3000s committed:
Update other messages

Updated StringId and category

Damon3000s committed:
Fix incorrectly named variable

Damon3000s committed:
Replace manual for loop converting paths to use existing function

Damon3000s committed:
Split the responsibility of FilesToSync for easier parsing

If FilesToSync is no longer a larger list that is stripped then the responsibility is clearer. FilesToDiff is the full list and the value we get back from the git operation is the subset of files we want to operate on (FilesToSync)

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,117 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		if (UneditableAssets.Num() != 0)
+		{
+			FString JoinedFiles = FString::Join(UneditableAssets, TEXT("\n"));
+			FMessageDialog::Open(EAppMsgType::Ok, FText::Format(LOCTEXT("FailedBranchSync", "Failed to sync file from branch {0} because the following files cannot be checked out \n{1}"), { FText::FromString(BranchName), FText::FromString(JoinedFiles) }));
```
Error dialog

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,117 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		{
+			FString JoinedFiles = FString::Join(UneditableAssets, TEXT("\n"));
+			FMessageDialog::Open(EAppMsgType::Ok, FText::Format(LOCTEXT("FailedBranchSync", "Failed to sync file from branch {0} because the following files cannot be checked out \n{1}"), { FText::FromString(BranchName), FText::FromString(JoinedFiles) }));
+			return;
```
Mid function return

Damon3000s replied:
Hate it, but not going to change it now.

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,117 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+		Provider.Execute(ISourceControlOperation::Create<FCheckOut>(), FilesToLock);
+
+		const bool bOperationSuccess = USourceControlHelpers::ApplyOperationAndReloadPackages(FilesToSync,
+			[&InPathToGitBinary, &InRepositoryRoot, &BranchName, &FilesToSync, &Provider, &AbsoluteFilePathToAsset](const TArray<FString>&)
```
What is this argument? Do we need to capture files to sync still?

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlUtils.cpp:
```diff
@@ -2529,6 +2533,117 @@ TSharedPtr<ISourceControlRevision, ESPMode::ThreadSafe> GetOriginRevisionOnBranc
+				{
+					PackagesToRevert.Add(AbsoluteFilePathToAsset[AbsoluteFilePath]);
+				}
+				auto RevertOperation = ISourceControlOperation::Create<FRevert>();
```
Can this be not auto?

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlModule.cpp:
```diff
@@ -199,6 +200,33 @@ void FGitSourceControlModule::CreateGitContentBrowserAssetMenu(FMenuBuilder& Men
+		return AssetData.AssetClassPath.GetAssetName() == WorldAssetName;
+	});
+	const bool bCanRevertToStatusBranch = !bAssetsContainsLevel;
+
```
no whitespace

Damon3000s commented on Source/GitSourceControl/Private/GitSourceControlModule.cpp:
```diff
@@ -199,6 +200,33 @@ void FGitSourceControlModule::CreateGitContentBrowserAssetMenu(FMenuBuilder& Men
+	{
+		return AssetData.AssetClassPath.GetAssetName() == WorldAssetName;
+	});
+	const bool bCanRevertToStatusBranch = !bAssetsContainsLevel;
```
whitespace above

Damon3000s committed:
Whitespace tweaks

Damon3000s committed:
Make failed dialog an error dialog

Damon3000s committed:
Fix usage of auto

Damon3000s committed:
Remove the revert step if git checkout fails

It wasn't working and decided it is more effort and risk than it is worth. If the command fails it is better to get the user to inspect the contents before decidding how to progress.

Damon3000s committed:
Fix tabbing

Damon3000s committed:
Always add the StatusRevert button, but disable if there is a map selected

Damon3000s committed:
Add basic cconfirmation before executing revert to status branch

Damon3000s committed:
Update variable naming for consistency

Damon3000s commented:
This PR has been superseded by a new integration branch (same code) https://github.com/TheThreeThousands/UEGitPlugin/pull/12/files
